### PR TITLE
Eagerly resolve discriminated unions and cache cases where we can't

### PR DESCRIPTION
**Pull request by adriangb**
Selected Reviewer: @dmontagu

### CODE REVIEW

File: pydantic/_internal/_generate_schema.py [hunk 1]
```diff
@@ -355,6 +356,22 @@ def _unknown_type_schema(self, obj: Any) -> CoreSchema:
...
+                discriminator,
+            )
+        except _discriminated_union.MissingDefinitionForUnionRef:
+            # defer until defs are resolved
+            _discriminated_union.set_discriminator(
+                schema,
+                discriminator,
+            )
+            schema.setdefault('metadata', {})[NEEDS_APPLY_DISCRIMINATED_UNION_METADATA_KEY] = True
+            self._needs_apply_discriminated_union = True
```
Comment (davidhewitt) [inline pydantic/_internal/_generate_schema.py:374]: I can't find any references to this field, maybe a hangover?

File: pydantic/_internal/_discriminated_union.py [hunk 2]
```diff
@@ -19,16 +29,23 @@ def set_discriminator(schema: CoreSchema, discriminator: Any) -> None:
 
 
 def apply_discriminators(schema: core_schema.CoreSchema) -> core_schema.CoreSchema:
-    definitions = collect_definitions(schema)
+    definitions: dict[str, CoreSchema] | None = None
 
     def inner(s: core_schema.CoreSchema, recurse: _core_utils.Recurse) -> core_schema.CoreSchema:
+        nonlocal definitions
+        if 'metadata' in s:
+            if s['metadata'].get(NEEDS_APPLY_DISCRIMINATED_UNION_METADATA_KEY, True) is False:
```
Comment (davidhewitt) [inline pydantic/_internal/_discriminated_union.py:37]: I can't see where this is ever set to false?

File: pydantic/_internal/_generate_schema.py (hunk 1 above)
Comment (adriangb) [reply pydantic/_internal/_generate_schema.py:374]: I think I pushed after you last refreshed

File: pydantic/_internal/_discriminated_union.py (hunk 2 above)
Comment (adriangb) [reply pydantic/_internal/_discriminated_union.py:37]: Same as above, refresh and you'll see it get set in `GenerateSchema._post_process_generated_schema`.

**Review by davidhewitt — APPROVED** [review]

**Pull request merged**